File: supabase/functions/chatbot-whatsapp-webhook/chatbot/main_handler.ts
import { createClient } from "https://esm.sh/@supabase/supabase-js@2"
import { sendSmartMessage, canUseWhatsApp } from "../../utility/communication.ts"
import { ConversationState, getConversationState, updateConversationState } from './conversation_state.ts'
import { getOrCreateCustomer } from './database_helpers.ts'
import { handleMainMenu } from './main_menu.ts'
import { handleOrdering } from './ordering.ts'
import { handleReviewFlow } from './review_system.ts'
import { formatMainMenu, validateMessageLength } from './message_formatter.ts'

const supabase = createClient(
  Deno.env.get('SUPABASE_URL') ?? '',
  Deno.env.get('SUPABASE_SERVICE_ROLE_KEY') ?? ''
)

// Always use this bot ID for all chats
const SYSTEM_BOT_ID = '46c4426a-9f5d-43d1-914c-d112deaf1d06'

export async function processMessage(
  phoneNumberId: string,
  from: string,
  messageId: string,
  messageBody: string,
  displayName?: string
): Promise<void> {
  try {
    // Normalize phone number
    const normalizedFrom = from.startsWith('+') ? from.substring(1) : from
    
    // Check if this is a consent message first
    const isConsentMessage = await handleConsentMessage(messageBody, normalizedFrom)
    if (isConsentMessage) {
      console.log('Consent processed successfully, comprehensive response sent')
      return
    }
    
    // Verify this is our bot
    const bot = await getBotFromPhoneNumber(phoneNumberId)
    if (!bot || bot.id !== SYSTEM_BOT_ID) {
      console.error('Message from unknown bot:', phoneNumberId)
      return
    }

    // Get or create customer
    const customer = await getOrCreateCustomer(normalizedFrom, displayName)
    if (!customer) {
      console.error('Failed to get/create customer')
      return
    }

    // Get or create chat
    const chat = await getOrCreateChat(customer.id)
    if (!chat) {
      console.error('Failed to get/create chat')
      return
    }

    // Save incoming message
    await saveMessage(chat.id, messageId, 'inbound', messageBody)

    // Check if this is a new conversation and handle recent bites
    const { data: previousMessages, error: messageError } = await supabase
      .from('messages')
      .select('id')
      .eq('chat_id', chat.id)
      .limit(1)

    const isNewConversation = !messageError && (!previousMessages || previousMessages.length === 0)

    if (isNewConversation) {
      console.log(`=== CHECKING RECENT BITES FOR NEW CONVERSATION ===`)
      console.log(`Customer ID: ${customer.id}, WhatsApp Number: ${normalizedFrom}`)
      
      const recentBites = await getRecentBitesByWhatsApp(normalizedFrom)
      console.log(`Found ${recentBites.length} recent bites for ${normalizedFrom}`)
      
      if (recentBites.length > 0) {
        console.log('Recent bites found:', recentBites.map(b => ({ id: b.id, order_number: b.order_number, status: b.status })))
        
        const updatesSent = await sendRecentBiteUpdates(normalizedFrom, recentBites)
        if (updatesSent) {
          await saveMessage(chat.id, '', 'outbound', 'Recent order updates')
          await updateChatLastMessage(chat.id, 'Recent order updates sent')
          return // Exit early since we've already responded
        }
      }
    }

    // Get conversation state
    const state = await getConversationState(chat.id)
    
    // Add debug logging for location messages
    if (messageBody.startsWith('LOCATION:')) {
      console.log('=== LOCATION MESSAGE DETECTED ===')
      console.log('Message body:', messageBody)
      console.log('Current conversation state:', state)
      console.log('Will route to handler for step:', state.step)
    }
    
    // Route to appropriate handler based on conversation state
    let responseMessage = ''
    
    try {
      switch (state.step) {
        case 'main_menu':
          responseMessage = await handleMainMenu(chat.id, customer.id, messageBody, state)
          break
          
        case 'new_order_warning':
          const { handleNewOrderWarning } = await import('./main_menu.ts')
          responseMessage = await handleNewOrderWarning(chat.id, customer.id, messageBody, state)
          break
          
        case 'order_type':
        case 'address_selection':
        case 'new_address':
        case 'store_selection':
        case 'store_search':
        case 'menu_display':
        case 'category_items':
        case 'item_details':
        case 'item_customization':
        case 'cart_view':
        case 'checkout':
        case 'tip_selection':
        case 'email_collection':
        case 'payment_method':
        case 'payment':
          responseMessage = await handleOrdering(chat.id, customer.id, messageBody, state)
          break
          
        case 'review_selection':
        case 'rating':
        case 'comment':
        case 'comment_write':
        case 'anon_selection':
        case 'completed':
          responseMessage = await handleReviewFlow(chat.id, customer.id, messageBody, state)
          break
          
        case 'address_list':
        case 'address_actions':
        case 'address_add':
        case 'address_added':
        case 'location_suggestions':
          const { handleAddressManagement } = await import('./address_management.ts')
          responseMessage = await handleAddressManagement(chat.id, customer.id, messageBody, state)
          break
          
        case 'profile_view':
        case 'profile_edit':
        case 'profile_field_edit':
          const { handleProfileManagement } = await import('./profile_management.ts')
          responseMessage = await handleProfileManagement(chat.id, customer.id, messageBody, state)
          break
          
        default:
          // Default to main menu
          responseMessage = formatMainMenu()
          await updateConversationState(chat.id, { step: 'main_menu' })
      }
    } catch (handlerError) {
      console.error('Error in message handler:', handlerError)
      console.error('Context:', { 
        step: state.step, 
        customerId: customer.id, 
        messageBody: messageBody.substring(0, 100) 
      })
      
      // Send generic error message to user instead of actual error
      responseMessage = `❌ *Service Temporarily Unavailable*\n\nWe're experiencing technical difficulties. Please try again in a few moments.\n\nIf the problem persists, please contact support.\n\n📱 *Powered by BeepBite.io*`
      
      // Reset to main menu to prevent user from being stuck
      try {
        await updateConversationState(chat.id, { step: 'main_menu' })
      } catch (resetError) {
        console.error('Error resetting conversation state:', resetError)
      }
    }

    // Send response
    await sendResponse(normalizedFrom, responseMessage, chat.id)

  } catch (error) {
    console.error('Error processing message:', error)
  }
}

export async function handleConsentMessage(messageBody: string, normalizedFrom: string): Promise<boolean> {
  // Regex to match CONSENT-{uuid} format - case insensitive
  const consentRegex = /CONSENT-([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})/i
  const match = messageBody.match(consentRegex)
  
  if (!match) {
    return false // Not a consent message
  }
  
  const biteId = match[1]
  console.log(`Consent message detected for bite ID: ${biteId} from ${normalizedFrom}`)
  
  try {
    // Get or create customer
    const customer = await getOrCreateCustomer(normalizedFrom)
    if (!customer) {
      console.error('Failed to get/create customer for consent')
      return false
    }

    // Get or create chat
    const chat = await getOrCreateChat(customer.id)
    if (!chat) {
      console.error('Failed to get/create chat for consent')
      return false
    }

    // Find the bite and verify it belongs to this customer
    const { data: bite, error: biteError } = await supabase
      .from('orders')
      .select(`
        *,
        customers!inner(whatsapp_number),
        locations(name)
      `)
      .eq('id', biteId)
      .eq('customers.whatsapp_number', normalizedFrom)
      .single()

    if (biteError || !bite) {
      console.error('Bite not found or does not belong to customer:', biteError)
      return false
    }

    console.log(`Valid consent for bite: ${bite.order_number}`)

    // Send consent confirmation and comprehensive bot introduction
    await sendConsentConfirmation(normalizedFrom, bite, chat.id)
    
    return true // Successfully processed consent
    
  } catch (error) {
    console.error('Error processing consent message:', error)
    return false
  }
}

export async function sendConsentConfirmation(whatsappNumber: string, connectedBite: any, chatId: string) {
  let message = `✅ *Thank you for confirming!*\n\n`
  message += `I found your order #${connectedBite.order_number} from ${connectedBite.locations?.name || 'the restaurant'}.\n\n`
  
  message += `🤖 *I'm your BeepBite assistant!*\n\n`
  message += `I can help you:\n`
  message += `📝 Track your orders\n`
  message += `⭐ Rate your experiences\n`
  message += `💬 Share feedback\n`
  message += `🔔 Get status updates\n\n`
  
  message += `Let me check if you have any orders to review...\n\n`
  message += `📱 *Powered by BeepBite.io*`

  console.log('Sending consent confirmation message')
  
  // Send the consent confirmation first
  const consentResult = await sendSmartMessage({
    whatsapp_number: whatsappNumber,
    message: message,
    subject: 'BeepBite.io - Welcome!'
  })

  if (consentResult.success && consentResult.method === 'whatsapp') {
    await saveMessage(chatId, '', 'outbound', message)
  }

  // Wait a moment then send the review flow
  setTimeout(async () => {
    try {
      // Import review functions
      const { getUnreviewedBites, formatBitesForReview, formatWelcomeMessage, getIncompleteBites } = await import('./review_system.ts')
      
      const unreviewedBites = await getUnreviewedBites(whatsappNumber)
      
      let followUpMessage = ''
      if (unreviewedBites.length > 0) {
        followUpMessage = formatBitesForReview(unreviewedBites, 0)
        // Update conversation state to review selection
        await updateConversationState(chatId, { step: 'review_selection', review_page: 0 })
      } else {
        const incompleteBites = await getIncompleteBites(whatsappNumber)
        followUpMessage = formatWelcomeMessage(incompleteBites, 0)
        // Update conversation state to main menu
        await updateConversationState(chatId, { step: 'main_menu', review_page: 0 })
      }

      const followUpResult = await sendSmartMessage({
        whatsapp_number: whatsappNumber,
        message: followUpMessage,
        subject: 'BeepBite.io - Your Orders'
      })

      if (followUpResult.success && followUpResult.method === 'whatsapp') {
        await saveMessage(chatId, '', 'outbound', followUpMessage)
      }

      await updateChatLastMessage(chatId, followUpMessage)
      
    } catch (error) {
      console.error('Error sending follow-up message:', error)
    }
  }, 2000) // 2 second delay
}

async function getRecentBitesByWhatsApp(whatsappNumber: string) {
  const { data: bites, error } = await supabase
    .from('orders')
    .select(`
      id,
      order_number,
      status,
      created_at,
      customers!inner(whatsapp_number),
      locations(name)
    `)
    .eq('customers.whatsapp_number', whatsappNumber)
    .gte('created_at', new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString()) // Last 24 hours
    .order('created_at', { ascending: false })
    .limit(5)

  if (error) {
    console.error('Error fetching recent bites:', error)
    return []
  }

  return bites || []
}

async function sendRecentBiteUpdates(whatsappNumber: string, recentBites: any[]): Promise<boolean> {
  try {
    let message = `🔔 *Order Updates*\n\n`
    message += `Here are your recent orders:\n\n`

    recentBites.forEach((bite, index) => {
      message += `*${index + 1}.* ${bite.locations?.name || 'Restaurant'}\n`
      message += `   Order #${bite.order_number}\n`
      message += `   Status: ${bite.status}\n`
      message += `   ${getTimeRemaining(bite.created_at)}\n\n`
    })

    message += `I'll help you track these and collect reviews when ready!\n\n`
    message += `📱 *Powered by BeepBite.io*`

    const result = await sendSmartMessage({
      whatsapp_number: whatsappNumber,
      message: message,
      subject: 'BeepBite.io - Order Updates'
    })

    return result.success
  } catch (error) {
    console.error('Error sending recent bite updates:', error)
    return false
  }
}

function getTimeRemaining(createdAt: string): string {
  const now = new Date()
  const created = new Date(createdAt)
  const diffMs = now.getTime() - created.getTime()
  const diffHours = Math.floor(diffMs / (1000 * 60 * 60))
  const diffMinutes = Math.floor((diffMs % (1000 * 60 * 60)) / (1000 * 60))
  
  if (diffHours > 0) {
    return `${diffHours}h ${diffMinutes}m ago`
  } else {
    return `${diffMinutes}m ago`
  }
}

async function sendResponse(whatsappNumber: string, message: string, chatId: string): Promise<void> {
  console.log('=== SENDING RESPONSE ===')
  
  // Validate message length before sending
  const validation = validateMessageLength(message)
  let messageToSend = message
  
  if (!validation.valid) {
    console.warn(`Message too long (${validation.length} chars), truncating to fit WhatsApp limit`)
    messageToSend = validation.truncated!
  }
  
  try {
    // Check if we can use WhatsApp for this customer
    const whatsappCheck = await canUseWhatsApp(whatsappNumber)
    console.log('WhatsApp capability check:', whatsappCheck)
    
    if (whatsappCheck.canUse) {
      // Can use WhatsApp - send directly
      const messageSent = await sendWhatsAppMessage(whatsappNumber, messageToSend)
      
      if (messageSent) {
        // Save outbound message
        await saveMessage(chatId, '', 'outbound', messageToSend)
        
        // Update chat with last message
        await updateChatLastMessage(chatId, messageToSend)
      } else {
        console.error('Failed to send WhatsApp message to:', whatsappNumber)
        
        // Try to send a fallback message
        try {
          const fallbackMessage = `❌ *Service Error*\n\nWe're having trouble sending your message. Please try again.\n\n📱 *Powered by BeepBite.io*`
          await sendWhatsAppMessage(whatsappNumber, fallbackMessage)
        } catch (fallbackError) {
          console.error('Failed to send fallback message:', fallbackError)
        }
      }
    } else {
      // Cannot use WhatsApp - use smart routing for fallbacks
      console.log('Cannot use WhatsApp for response - using smart routing')
      
      const smartResult = await sendSmartMessage({
        whatsapp_number: whatsappNumber,
        message: messageToSend,
        subject: 'BeepBite.io - Restaurant Chat'
      })
      
      console.log('Smart message result:', smartResult)
      
      // Still update chat state regardless of send method
      await updateChatLastMessage(chatId, messageToSend)
      
      // Only save to messages table if it was sent via WhatsApp
      if (smartResult.method === 'whatsapp') {
        await saveMessage(chatId, '', 'outbound', messageToSend)
      }
    }
  } catch (error) {
    console.error('Error in sendResponse:', error)
    console.error('Failed message details:', {
      whatsappNumber,
      messageLength: messageToSend.length,
      chatId
    })
    
    // Don't expose internal errors to users - they're already handled in processMessage
  }
}

async function getBotFromPhoneNumber(phoneNumberId: string) {
  const { data: bot, error } = await supabase
    .from('bots')
    .select('*')
    .eq('whatsapp_phone_number_id', phoneNumberId)
    .eq('is_active', true)
    .single()

  if (error) {
    console.error('Error fetching bot:', error)
    return null
  }

  return bot
}

async function getOrCreateChat(customerId: string, locationId?: string) {
  // Check if active chat exists for this customer with our bot
  const { data: existingChat, error: fetchError } = await supabase
    .from('chats')
    .select('*')
    .eq('bot_id', SYSTEM_BOT_ID)
    .eq('customer_id', customerId)
    .eq('status', 'active')
    .single()

  if (fetchError && fetchError.code !== 'PGRST116') {
    console.error('Error fetching chat:', fetchError)
    return null
  }

  if (existingChat) {
    // Update location_id if provided and different
    if (locationId && existingChat.location_id !== locationId) {
      const { error: updateError } = await supabase
        .from('chats')
        .update({ location_id: locationId })
        .eq('id', existingChat.id)
        
      if (updateError) {
        console.error('Error updating chat location:', updateError)
      }
    }
    return existingChat
  }

  // Create new chat
  const { data: newChat, error: createError } = await supabase
    .from('chats')
    .insert({
      bot_id: SYSTEM_BOT_ID,
      customer_id: customerId,
      location_id: locationId || null, // Can be null initially
      status: 'active',
      conversation_state: { step: 'main_menu' },
      bot_active: true,
      last_message_at: new Date().toISOString()
    })
    .select()
    .single()

  if (createError) {
    console.error('Error creating chat:', createError)
    return null
  }

  return newChat
}

async function saveMessage(chatId: string, whatsappMessageId: string, direction: 'inbound' | 'outbound', content: string) {
  const { error } = await supabase
    .from('messages')
    .insert({
      chat_id: chatId,
      whatsapp_message_id: whatsappMessageId,
      direction: direction,
      message_type: 'text',
      content: content,
      created_at: new Date().toISOString()
    })

  if (error) {
    console.error('Error saving message:', error)
  }
}

async function updateChatLastMessage(chatId: string, message: string) {
  const { error } = await supabase
    .from('chats')
    .update({
      last_message_at: new Date().toISOString(),
      last_message_preview: message.substring(0, 100)
    })
    .eq('id', chatId)

  if (error) {
    console.error('Error updating chat last message:', error)
  }
}

export async function sendWhatsAppMessage(to: string, message: string): Promise<boolean> {
  try {
    const accessToken = Deno.env.get('WHATSAPP_API_TOKEN')
    
    if (!accessToken) {
      console.error('WhatsApp access token not configured')
      return false
    }

    // Get phone number ID from bot record instead of environment variable
    const { data: bot, error: botError } = await supabase
      .from('bots')
      .select('whatsapp_phone_number_id')
      .eq('id', SYSTEM_BOT_ID)
      .eq('is_active', true)
      .single()

    if (botError || !bot?.whatsapp_phone_number_id) {
      console.error('Bot not found or phone number ID not configured:', botError)
      return false
    }

    const phoneNumberId = bot.whatsapp_phone_number_id

    const response = await fetch(`https://graph.facebook.com/v17.0/${phoneNumberId}/messages`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${accessToken}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        messaging_product: 'whatsapp',
        recipient_type: 'individual', // Can send to individual WhatsApp numbers
        to: to,
        type: 'text',
        text: {
          preview_url: false,
          body: message
        }
      })
    })

    if (!response.ok) {
      const errorData = await response.json()
      console.error('WhatsApp API error:', errorData)
      return false
    }

    console.log(`WhatsApp message sent successfully to ${to} via phone number ID: ${phoneNumberId}`)
    return true
  } catch (error) {
    console.error('Error sending WhatsApp message:', error)
    return false
  }
} 

File: backend/internal/chatbot/main_handler.go
package chatbot

import (
	"context"
	"errors"
	"fmt"
	"log"
	"regexp"
	"strings"
	"time"

	"github.com/jackc/pgx/v5"
)

var consentRegex = regexp.MustCompile(`(?i)CONSENT-([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})`)

// ProcessMessage is the main entrypoint called by the webhook handler.
func (s *Service) ProcessMessage(ctx context.Context, phoneNumberID, from, messageID, messageBody, displayName string) error {
	defer func() {
		if r := recover(); r != nil {
			log.Printf("Panic in ProcessMessage: %v", r)
		}
	}()

	normalizedFrom := from
	if strings.HasPrefix(normalizedFrom, "+") {
		normalizedFrom = normalizedFrom[1:]
	}

	// Consent message shortcut
	isConsent, err := s.handleConsentMessage(ctx, messageBody, normalizedFrom)
	if err != nil {
		log.Printf("Error handling consent: %v", err)
	}
	if isConsent {
		log.Printf("Consent processed successfully, comprehensive response sent")
		return nil
	}

	bot := s.getBotFromPhoneNumber(ctx, phoneNumberID)
	if bot == nil || bot.ID != SystemBotID {
		log.Printf("Message from unknown bot: %s", phoneNumberID)
		return nil
	}

	customer := s.getOrCreateCustomer(ctx, normalizedFrom, displayName)
	if customer == nil {
		log.Printf("Failed to get/create customer")
		return nil
	}

	chat := s.getOrCreateChat(ctx, customer.ID, "")
	if chat == nil {
		log.Printf("Failed to get/create chat")
		return nil
	}

	s.saveMessage(ctx, chat.ID, messageID, "inbound", messageBody)

	// Check new conversation + recent bites
	var msgCount int
	err = s.pool.QueryRow(ctx,
		`SELECT COUNT(*) FROM (SELECT id FROM messages WHERE chat_id = $1 LIMIT 1) AS m`,
		chat.ID,
	).Scan(&msgCount)
	isNewConversation := err == nil && msgCount <= 1 // we just inserted one

	if isNewConversation {
		log.Printf("=== CHECKING RECENT BITES FOR NEW CONVERSATION ===")
		recent := s.getRecentBitesByWhatsApp(ctx, normalizedFrom)
		if len(recent) > 0 {
			if s.sendRecentBiteUpdates(normalizedFrom, recent) {
				s.saveMessage(ctx, chat.ID, "", "outbound", "Recent order updates")
				s.updateChatLastMessage(ctx, chat.ID, "Recent order updates sent")
				return nil
			}
		}
	}

	state := s.getConversationState(ctx, chat.ID)

	if strings.HasPrefix(messageBody, "LOCATION:") {
		log.Printf("=== LOCATION MESSAGE DETECTED === body=%s state.step=%s", messageBody, state.Step)
	}

	responseMessage := s.routeMessage(ctx, chat.ID, customer.ID, messageBody, state)

	s.sendResponse(ctx, normalizedFrom, responseMessage, chat.ID)
	return nil
}

func (s *Service) routeMessage(ctx context.Context, chatID, customerID, messageBody string, state ConversationState) (resp string) {
	defer func() {
		if r := recover(); r != nil {
			log.Printf("Panic in handler (step=%s): %v", state.Step, r)
			resp = "❌ *Service Temporarily Unavailable*\n\nWe're experiencing technical difficulties. Please try again in a few moments.\n\nIf the problem persists, please contact support.\n\n📱 *Powered by BeepBite.io*"
			s.updateConversationState(ctx, chatID, ConversationState{Step: "main_menu"})
		}
	}()

	switch state.Step {
	case "main_menu":
		return s.handleMainMenu(ctx, chatID, customerID, messageBody, state)

	case "new_order_warning":
		return s.handleNewOrderWarning(ctx, chatID, customerID, messageBody, state)

	case "order_type", "address_selection", "new_address", "store_selection",
		"store_search", "menu_display", "category_items", "item_details",
		"item_customization", "cart_view", "checkout", "tip_selection",
		"email_collection", "payment_method", "payment":
		return s.handleOrdering(ctx, chatID, customerID, messageBody, state)

	case "review_selection", "rating", "comment", "comment_write", "anon_selection", "completed":
		return s.handleReviewFlow(ctx, chatID, customerID, messageBody, state)

	case "address_list", "address_actions", "address_add", "address_added", "location_suggestions":
		return s.handleAddressManagement(ctx, chatID, customerID, messageBody, state)

	case "profile_view", "profile_edit", "profile_field_edit":
		return s.handleProfileManagement(ctx, chatID, customerID, messageBody, state)

	case "billing_list", "billing_actions", "billing_add_email_check", "billing_add", "billing_added":
		return s.handleBillingManagement(ctx, chatID, customerID, messageBody, state)

	default:
		s.updateConversationState(ctx, chatID, ConversationState{Step: "main_menu"})
		return formatMainMenu("", 0, 0, "")
	}
}

func (s *Service) handleConsentMessage(ctx context.Context, messageBody, normalizedFrom string) (bool, error) {
	match := consentRegex.FindStringSubmatch(messageBody)
	if len(match) < 2 {
		return false, nil
	}
	biteID := match[1]
	log.Printf("Consent message detected for bite ID: %s from %s", biteID, normalizedFrom)

	customer := s.getOrCreateCustomer(ctx, normalizedFrom, "")
	if customer == nil {
		return false, errors.New("failed to get/create customer for consent")
	}
	chat := s.getOrCreateChat(ctx, customer.ID, "")
	if chat == nil {
		return false, errors.New("failed to get/create chat for consent")
	}

	// Validate bite belongs to customer
	var orderNumber string
	var locationName *string
	err := s.pool.QueryRow(ctx,
		`SELECT o.order_number, l.name
		 FROM orders o
		 JOIN customers c ON o.customer_id = c.id
		 LEFT JOIN locations l ON o.location_id = l.id
		 WHERE o.id = $1 AND c.whatsapp_number = $2`,
		biteID, normalizedFrom,
	).Scan(&orderNumber, &locationName)
	if err != nil {
		log.Printf("Bite not found or does not belong to customer: %v", err)
		return false, nil
	}

	log.Printf("Valid consent for bite: %s", orderNumber)
	s.sendConsentConfirmation(ctx, normalizedFrom, orderNumber, locationName, chat.ID)
	return true, nil
}

func (s *Service) sendConsentConfirmation(ctx context.Context, whatsappNumber, orderNumber string, locationName *string, chatID string) {
	loc := "the restaurant"
	if locationName != nil && *locationName != "" {
		loc = *locationName
	}
	var b strings.Builder
	b.WriteString("✅ *Thank you for confirming!*\n\n")
	fmt.Fprintf(&b, "I found your order #%s from %s.\n\n", orderNumber, loc)
	b.WriteString("🤖 *I'm your BeepBite assistant!*\n\n")
	b.WriteString("I can help you:\n")
	b.WriteString("📝 Track your orders\n")
	b.WriteString("⭐ Rate your experiences\n")
	b.WriteString("💬 Share feedback\n")
	b.WriteString("🔔 Get status updates\n\n")
	b.WriteString("Let me check if you have any orders to review...\n\n")
	b.WriteString("📱 *Powered by BeepBite.io*")
	message := b.String()

	log.Printf("Sending consent confirmation message")
	if s.sendWhatsAppMessage(whatsappNumber, message) {
		s.saveMessage(ctx, chatID, "", "outbound", message)
	}

	// Send follow-up inline (TS uses a 2s timeout; in Go we just send directly).
	unreviewed := s.getUnreviewedBites(ctx, whatsappNumber)
	var followUp string
	if len(unreviewed) > 0 {
		followUp = formatBitesForReview(unreviewed, 0)
		newState := ConversationState{Step: "review_selection", ReviewPage: 0}
		s.updateConversationState(ctx, chatID, newState)
	} else {
		incomplete := s.getIncompleteBites(ctx, whatsappNumber)
		followUp = formatWelcomeMessage(incomplete, 0)
		newState := ConversationState{Step: "main_menu", ReviewPage: 0}
		s.updateConversationState(ctx, chatID, newState)
	}

	if s.sendWhatsAppMessage(whatsappNumber, followUp) {
		s.saveMessage(ctx, chatID, "", "outbound", followUp)
	}
	s.updateChatLastMessage(ctx, chatID, followUp)
}

func (s *Service) getRecentBitesByWhatsApp(ctx context.Context, whatsappNumber string) []reviewBite {
	since := time.Now().Add(-24 * time.Hour)
	rows, err := s.pool.Query(ctx,
		`SELECT o.id, o.order_number, o.status, o.created_at, l.name
		 FROM orders o
		 JOIN customers c ON o.customer_id = c.id
		 LEFT JOIN locations l ON o.location_id = l.id
		 WHERE c.whatsapp_number = $1 AND o.created_at >= $2
		 ORDER BY o.created_at DESC
		 LIMIT 5`,
		whatsappNumber, since,
	)
	if err != nil {
		log.Printf("Error fetching recent bites: %v", err)
		return nil
	}
	defer rows.Close()
	var bites []reviewBite
	for rows.Next() {
		b := reviewBite{}
		var name *string
		if err := rows.Scan(&b.ID, &b.OrderNumber, &b.Status, &b.CreatedAt, &name); err != nil {
			log.Printf("Error scanning recent bite: %v", err)
			continue
		}
		if name != nil {
			b.LocationName = *name
		}
		bites = append(bites, b)
	}
	return bites
}

func (s *Service) sendRecentBiteUpdates(whatsappNumber string, recentBites []reviewBite) bool {
	var b strings.Builder
	b.WriteString("🔔 *Order Updates*\n\n")
	b.WriteString("Here are your recent orders:\n\n")
	for i, bite := range recentBites {
		loc := bite.LocationName
		if loc == "" {
			loc = "Restaurant"
		}
		fmt.Fprintf(&b, "*%d.* %s\n", i+1, loc)
		fmt.Fprintf(&b, "   Order #%s\n", bite.OrderNumber)
		fmt.Fprintf(&b, "   Status: %s\n", bite.Status)
		fmt.Fprintf(&b, "   %s\n\n", getTimeRemaining(bite.CreatedAt))
	}
	b.WriteString("I'll help you track these and collect reviews when ready!\n\n")
	b.WriteString("📱 *Powered by BeepBite.io*")
	return s.sendWhatsAppMessage(whatsappNumber, b.String())
}

func (s *Service) sendResponse(ctx context.Context, whatsappNumber, message, chatID string) {
	log.Printf("=== SENDING RESPONSE ===")
	validation := validateMessageLength(message)
	toSend := message
	if !validation.Valid {
		log.Printf("Message too long (%d chars), truncating", validation.Length)
		toSend = validation.Truncated
	}

	if !canUseWhatsApp(whatsappNumber) {
		// Fallback path — no other transport available in this port.
		s.updateChatLastMessage(ctx, chatID, toSend)
		return
	}

	if s.sendWhatsAppMessage(whatsappNumber, toSend) {
		s.saveMessage(ctx, chatID, "", "outbound", toSend)
		s.updateChatLastMessage(ctx, chatID, toSend)
		return
	}

	log.Printf("Failed to send WhatsApp message to: %s", whatsappNumber)
	fallback := "❌ *Service Error*\n\nWe're having trouble sending your message. Please try again.\n\n📱 *Powered by BeepBite.io*"
	_ = s.sendWhatsAppMessage(whatsappNumber, fallback)
}

type botRecord struct {
	ID                    string
	WhatsappPhoneNumberID string
}

func (s *Service) getBotFromPhoneNumber(ctx context.Context, phoneNumberID string) *botRecord {
	b := &botRecord{}
	err := s.pool.QueryRow(ctx,
		`SELECT id, whatsapp_phone_number_id FROM bots
		 WHERE whatsapp_phone_number_id = $1 AND is_active = true`,
		phoneNumberID,
	).Scan(&b.ID, &b.WhatsappPhoneNumberID)
	if err != nil {
		log.Printf("Error fetching bot: %v", err)
		return nil
	}
	return b
}

type chatRecord struct {
	ID         string
	LocationID *string
}

func (s *Service) getOrCreateChat(ctx context.Context, customerID, locationID string) *chatRecord {
	existing := &chatRecord{}
	err := s.pool.QueryRow(ctx,
		`SELECT id, location_id FROM chats
		 WHERE bot_id = $1 AND customer_id = $2 AND status = 'active'
		 LIMIT 1`,
		SystemBotID, customerID,
	).Scan(&existing.ID, &existing.LocationID)

	if err == nil {
		if locationID != "" {
			current := ""
			if existing.LocationID != nil {
				current = *existing.LocationID
			}
			if current != locationID {
				_, updErr := s.pool.Exec(ctx,
					`UPDATE chats SET location_id = $1 WHERE id = $2`,
					locationID, existing.ID,
				)
				if updErr != nil {
					log.Printf("Error updating chat location: %v", updErr)
				}
			}
		}
		return existing
	}
	if !errors.Is(err, pgx.ErrNoRows) {
		log.Printf("Error fetching chat: %v", err)
	}

	// Create new chat
	var locArg interface{}
	if locationID != "" {
		locArg = locationID
	} else {
		locArg = nil
	}
	newChat := &chatRecord{}
	err = s.pool.QueryRow(ctx,
		`INSERT INTO chats
		   (bot_id, customer_id, location_id, status, conversation_state, last_message_at)
		 VALUES ($1, $2, $3, 'active', $4::jsonb, NOW())
		 RETURNING id, location_id`,
		SystemBotID, customerID, locArg, `{"step":"main_menu"}`,
	).Scan(&newChat.ID, &newChat.LocationID)
	if err != nil {
		log.Printf("Error creating chat: %v", err)
		return nil
	}
	return newChat
}

func (s *Service) saveMessage(ctx context.Context, chatID, whatsappMessageID, direction, content string) {
	var msgIDArg interface{}
	if whatsappMessageID != "" {
		msgIDArg = whatsappMessageID
	} else {
		msgIDArg = nil
	}
	_, err := s.pool.Exec(ctx,
		`INSERT INTO messages (chat_id, whatsapp_message_id, direction, message_type, content, created_at)
		 VALUES ($1, $2, $3, 'text', $4, NOW())`,
		chatID, msgIDArg, direction, content,
	)
	if err != nil {
		log.Printf("Error saving message: %v", err)
	}
}

func (s *Service) updateChatLastMessage(ctx context.Context, chatID, message string) {
	preview := message
	if len(preview) > 100 {
		preview = preview[:100]
	}
	_, err := s.pool.Exec(ctx,
		`UPDATE chats SET last_message_at = NOW(), last_message_preview = $1 WHERE id = $2`,
		preview, chatID,
	)
	if err != nil {
		log.Printf("Error updating chat last message: %v", err)
	}
}

func (s *Service) sendWhatsAppMessage(to, message string) bool {
	if s.wa == nil {
		log.Printf("WhatsApp client not configured")
		return false
	}
	_, err := s.wa.SendText(to, message, false)
	if err != nil {
		log.Printf("Error sending WhatsApp message: %v", err)
		return false
	}
	log.Printf("WhatsApp message sent successfully to %s", to)
	return true
}
